src/main/java/search/ingester/Processor.java

```diff
@@ -0,0 +1,125 @@
+
+import search.ingester.models.Document;
+import search.ingester.models.Message;
+
+public class Processor {
+
+    private ElasticService elasticService;
+    private FileParser fileParser;
+
+    public Processor(ElasticService elasticService, FileParser fileParser) {
+      this.elasticService = elasticService;
+      this.fileParser = fileParser;
+    }
+
+    public void process(Message m) throws IOException {
+
+        switch (m.getVerb()) {
+            case "upsert": processUpsert(m); break;
+            case "delete": processDelete(m); break;
+            case "spike":  processSpike(m);  break;
+            default: throw new RuntimeException(
+                String.format("Expected verb to be 'upsert' or 'delete' but got %s", m.getVerb()));
+        }        
+    }
+
+    void processUpsert(Message m) throws IOException {
+        
+        Document doc = m.getDocument();
+
+        System.out.println(":: Upserting doc " + doc.getId() + " for site " + doc.getSite() + " in index " + m.getIndex() + " ::");
+
+        deleteDatahubResourcesIfNecessary(m.getIndex(), doc);
+        extractContentFromFileBase64IfNecessary(doc);        
+        validateDocument(doc);
+        DocumentTweaker.setContentTruncatedField(doc);
+        elasticService.putDocument(m.getIndex(), doc);
+        upsertDatahubResourcesIfAny(m.getIndex(), doc, m.getResources());
+    }
+    
+    void deleteDatahubResourcesIfNecessary(String index, Document doc) throws IOException {
+
+        // if this is a datahub doc, delete any existing resources
+        if (doc.getSite().equals("datahub")) {
+            elasticService.deleteByParentId(index, doc.getId());
+        }
+    }
+
+    void extractContentFromFileBase64IfNecessary(Document doc) {
+        
+        // if this doc represents a "file" (e.g. a PDF) then it will have a file_base64 field
+        // which we need to extract into the content field etc.
+        if (doc.getFileBase64() != null) {
+            try {
+                // note this function mutates its argument (and returns it for good measure!)
+                doc = fileParser.parseFile(doc);
+            } catch (Exception err) {
+                throw new RuntimeException(err);
+            }
+        }
+    }
+
+    void validateDocument(Document doc) {
+
+        ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
+        Validator validator = factory.getValidator();
+        Set<ConstraintViolation<Document>> violations = validator.validate(doc);
+    
+        if (violations.size() > 0) {
+            throw new RuntimeException(
+                    violations.stream()
+                            .map(violation -> violation.getPropertyPath().toString() + ": " + violation.getMessage())
+                            .collect(Collectors.joining("\n")));
+        }
+    }
+
+    void upsertDatahubResourcesIfAny(String index, Document doc, List<Document> resources) throws IOException {
```

Comment:
This function doesn't seem to actually extract PDF content in any way, its looks like it would just dump the entire record with base64 encoding on the search index?

Also doesn't set content truncated field

Reply:
Oops! Steps were missing... put them back - great catch!